### PR TITLE
[Core] Transfer distributed matrix to a serial one

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**📝 Description**
A constructor for serial CSR by dictionaries in the form 
({I,J}:value)

and a function to transfer a distributed matrix to its serial counterpart (useful for the coarse solver of DD preconditioners)

Please mark the PR with appropriate tags: 
Feature

**🆕 Changelog**

### CODE REVIEW

File: kratos/mpi/tests/test_distributed_sparse_matrices.py [hunk 1]
```diff
@@ -167,7 +167,19 @@ def test_matrix_construction(self):
...
+        Aserial = A.ToSerialCSR(target_rank)
+        if(my_rank == 0):
+            y_serial = KratosMultiphysics.Vector(Aserial.Size1())
+            y_serial.fill(1.0)
+            b_serial = KratosMultiphysics.Vector(Aserial.Size1())
+            b_serial.fill(0.0)
+            Aserial.SpMV(y_serial, b_serial)
+            
+            for i in range(y_serial.Size()):
+                self.assertEqual(b_serial[i], reference_spmv_res[i], 1e-14)
```
Comment (philbucher) [inline kratos/mpi/tests/test_distributed_sparse_matrices.py]:
```suggestion
            self.assertVectorAlmostEqual(b_serial, reference_spmv_res)

```

File: kratos/containers/csr_matrix.h [hunk 2]
```diff
@@ -108,6 +108,26 @@ class CsrMatrix final
         SetValue(0.0);
     }
 
+    CsrMatrix(const MatrixMapType& matrix_map)
+    {
+        SparseGraph<TIndexType> Agraph;
+        for(const auto item : matrix_map){
```
Comment (loumalouomega) [inline kratos/containers/csr_matrix.h]:
```suggestion
        for(const auto& item : matrix_map){
```
?, I don't know if this is a gain

File: kratos/containers/distributed_csr_matrix.h [hunk 3]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
         return value_map;
     }
 
+    typename CsrMatrix<TDataType,TIndexType>::Pointer ToSerialCSR(MpiIndexType target_rank=0) const
+    {
+        //flatten all data (both indices and values) into a single vector of doubles
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h]:
```suggestion
        // Flatten all data (both indices and values) into a single vector of doubles
```

File: kratos/containers/distributed_csr_matrix.h [hunk 4]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+        std::vector<double> tmp_data;
+        tmp_data.reserve(GetDiagonalBlock().nnz()*3 + GetOffDiagonalBlock().nnz()*3);
+        for(unsigned int i=0; i<GetDiagonalBlock().size1(); ++i){
+            IndexType row_begin = GetDiagonalBlock().index1_data()[i];
+            IndexType row_end   = GetDiagonalBlock().index1_data()[i+1];
+            for(IndexType k = row_begin; k < row_end; ++k)
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:703]:
```suggestion
            for(IndexType k = row_begin; k < row_end; ++k) {
```

File: kratos/containers/distributed_csr_matrix.h [hunk 5]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+            {
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:704]:
```suggestion
```

File: kratos/containers/distributed_csr_matrix.h [hunk 6]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                IndexType j = GetDiagonalBlock().index2_data()[k];
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:705]:
```suggestion
                const IndexType j = GetDiagonalBlock().index2_data()[k];
```

File: kratos/containers/distributed_csr_matrix.h [hunk 7]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                TDataType v = GetDiagonalBlock().value_data()[k];
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:706]:
```suggestion
                const TDataType v = GetDiagonalBlock().value_data()[k];
```

File: kratos/containers/distributed_csr_matrix.h [hunk 8]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                tmp_data.push_back(GetRowNumbering().GlobalId(i));
+                tmp_data.push_back(GetColNumbering().GlobalId(j));
+                tmp_data.push_back(v);
+            }
+        }
+        for(unsigned int i=0; i<GetOffDiagonalBlock().size1(); ++i){
+            IndexType row_begin = GetOffDiagonalBlock().index1_data()[i];
+            IndexType row_end   = GetOffDiagonalBlock().index1_data()[i+1];
+            for(IndexType k = row_begin; k < row_end; ++k)
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:715]:
```suggestion
            for(IndexType k = row_begin; k < row_end; ++k)  {
```

File: kratos/containers/distributed_csr_matrix.h [hunk 9]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+            {
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:716]:
```suggestion
```

File: kratos/containers/distributed_csr_matrix.h [hunk 10]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                IndexType j = GetOffDiagonalBlock().index2_data()[k];
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:717]:
```suggestion
                const IndexType j = GetOffDiagonalBlock().index2_data()[k]; 
```

File: kratos/containers/distributed_csr_matrix.h [hunk 11]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                tmp_data.push_back(v);
+            }
+        }
+        for(unsigned int i=0; i<GetOffDiagonalBlock().size1(); ++i){
+            IndexType row_begin = GetOffDiagonalBlock().index1_data()[i];
+            IndexType row_end   = GetOffDiagonalBlock().index1_data()[i+1];
+            for(IndexType k = row_begin; k < row_end; ++k)
+            {
+                IndexType j = GetOffDiagonalBlock().index2_data()[k]; 
+                TDataType v = GetOffDiagonalBlock().value_data()[k];
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:718]:
```suggestion
                const TDataType v = GetOffDiagonalBlock().value_data()[k];
```

File: kratos/containers/distributed_csr_matrix.h [hunk 12]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+        typename SparseContiguousRowGraph<TIndexType>::UniquePointer p_csr_graph;
+        typename CsrMatrix<TDataType,TIndexType>::Pointer p_csr_output;
+
+        if(my_rank==target_rank){
+            p_csr_graph = std::move( Kratos::make_unique<SparseContiguousRowGraph<TIndexType>>(GetRowNumbering().Size()) );
+
+            for(int i_proc=0; i_proc<num_processors; ++i_proc){
+                const auto& data = collected_data[i_proc];
+                for(IndexType i=0; i<data.size(); i+=3){
+                    IndexType I = static_cast<IndexType>(data[i]);
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:739]:
```suggestion
                    const IndexType I = static_cast<IndexType>(data[i]);
```

File: kratos/containers/distributed_csr_matrix.h [hunk 13]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                    IndexType J = static_cast<IndexType>(data[i+1]);
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h:740]:
```suggestion
                    const IndexType J = static_cast<IndexType>(data[i+1]);
```

File: kratos/containers/distributed_csr_matrix.h [hunk 14]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                TDataType v = GetOffDiagonalBlock().value_data()[k];
+                tmp_data.push_back(GetRowNumbering().GlobalId(i));
+                tmp_data.push_back(mOffDiagonalGlobalIds[j]);
+                tmp_data.push_back(v);
+            }
+        }
+
+        auto collected_data = GetComm().Gatherv(tmp_data,target_rank);
+
+        MpiIndexType num_processors = GetComm().Size();
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h]:
```suggestion
        const MpiIndexType num_processors = GetComm().Size();
```

File: kratos/containers/distributed_csr_matrix.h [hunk 15]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+        MpiIndexType my_rank = GetComm().Rank();
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h]:
```suggestion
        const MpiIndexType my_rank = GetComm().Rank();
```

File: kratos/containers/distributed_csr_matrix.h [hunk 16]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+            p_csr_graph->Finalize();
+        }
+
+        if(my_rank==target_rank){
+            p_csr_output = Kratos::make_shared<CsrMatrix<TDataType,TIndexType>>(*p_csr_graph);
+            p_csr_output->BeginAssemble();
+            for(int i_proc=0; i_proc<num_processors; ++i_proc){
+                const auto& data = collected_data[i_proc];
+                for(IndexType i=0; i<data.size(); i+=3){
+                    IndexType I = static_cast<IndexType>(data[i]);
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h]:
```suggestion
                    const IndexType I = static_cast<IndexType>(data[i]);
```

File: kratos/containers/distributed_csr_matrix.h [hunk 17]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                    IndexType J = static_cast<IndexType>(data[i+1]);
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h]:
```suggestion
                    const IndexType J = static_cast<IndexType>(data[i+1]);
```

File: kratos/containers/distributed_csr_matrix.h [hunk 18]
```diff
@@ -692,6 +692,76 @@ class DistributedCsrMatrix final
...
+                    double    v = data[i+2];
```
Comment (loumalouomega) [inline kratos/containers/distributed_csr_matrix.h]:
```suggestion
                    const double    v = data[i+2];
```

**Review by loumalouomega — COMMENTED** [review]
Many const

**Review by philbucher — COMMENTED** [review]
wow this is very useful!

File: kratos/containers/csr_matrix.h (hunk 2 above)
Comment (RiccardoRossi) [reply kratos/containers/csr_matrix.h]: good catch!

**Review by RiccardoRossi — COMMENTED** [review]
in principle i should have addressed all the comments

**Review by philbucher — APPROVED** [review]
cool :+1:

**Pull request merged**